src/test/java/com/snowflake/kafka/connector/internal/streaming/StreamingClientProviderTest.java

```diff
@@ -0,0 +1,363 @@
+    Mockito.verify(streamingIngestClient, Mockito.times(1)).isClosed();
+    Mockito.verify(streamingIngestClient, Mockito.times(2)).getName();
+  }
+
+  @Test
+  public void testCloseInvalidClient() throws Exception {
+    // inject invalid client
+    SnowflakeStreamingIngestClient streamingIngestClient =
+        Mockito.mock(SnowflakeStreamingIngestClient.class);
+    Mockito.when(streamingIngestClient.isClosed()).thenReturn(true);
+    StreamingClientProvider injectedProvider =
+        injectStreamingClientProviderForTests(1, this.connectorConfig, streamingIngestClient);
+
+    // try closing client
+    injectedProvider.closeClient();
+
+    // verify didn't call close
+    Mockito.verify(streamingIngestClient, Mockito.times(0)).close();
+  }
+
+  @Test
+  public void testCloseClientWithVariousExceptions() throws Exception {
+    List<Exception> exceptionsToTest = new ArrayList<>();
+
+    Exception nullMessageEx = new Exception();
+    exceptionsToTest.add(nullMessageEx);
+
+    Exception nullCauseEx = new Exception("nullCauseEx");
+    nullCauseEx.initCause(null);
+    exceptionsToTest.add(nullCauseEx);
+
+    Exception stacktraceEx = new Exception("stacktraceEx");
+    stacktraceEx.initCause(new Exception("cause"));
+    stacktraceEx.getCause().setStackTrace(new StackTraceElement[0]);
+    exceptionsToTest.add(stacktraceEx);
+
+    for (Exception ex : exceptionsToTest) {
+      this.testCloseClientWithExceptionRunner(ex);
+    }
+  }
+
+  @Test
+  public void testMultiThreadGetEnabledParam() {
+    String clientName = "clientName";
+    int clientId = 0;
+
+    // setup
+    this.connectorConfig.put(Utils.NAME, clientName);
+    SnowflakeStreamingIngestClient streamingIngestClient =
+        Mockito.mock(SnowflakeStreamingIngestClient.class);
+    StreamingClientProvider injectedProvider =
+        injectStreamingClientProviderForTests(
+            clientId, this.connectorConfig, streamingIngestClient);
+    GetClientRunnable getClientRunnable1 =
+        new GetClientRunnable(injectedProvider, this.connectorConfig, "getClientRunnable1");
+    GetClientRunnable getClientRunnable2 =
+        new GetClientRunnable(injectedProvider, this.connectorConfig, "getClientRunnable2");
+    GetClientRunnable getClientRunnable3 =
+        new GetClientRunnable(injectedProvider, this.connectorConfig, "getClientRunnable3");
+
+    // get client on multiple threads
+    getClientRunnable1.start();
+    getClientRunnable2.start();
+    getClientRunnable3.start();
+
+    // verify same client
+    SnowflakeStreamingIngestClient client1 = getClientRunnable1.getClient();
+    SnowflakeStreamingIngestClient client2 = getClientRunnable2.getClient();
+    SnowflakeStreamingIngestClient client3 = getClientRunnable3.getClient();
+
+    assert client1.getName().contains(clientName);
+    assert client1.getName().equals(client2.getName());
+    assert client2.getName().equals(client3.getName());
+  }
+
+  @Test
+  public void testMultiThreadGetDisabledParam() {
```

Comment:
can you call closeClient API here in case optimization is false? I think it will throw NPE.